### ISSUE TITLE
SkinnedMesh: Incorrect bounding box and culling

### ISSUE THREAD

**Issue by soadzoor** [mrdoob/three.js#14499]
##### Description of the problem

Hi,

I've been playing around with a model for a while, and noticed that if I load it in three.js and zoom closely to the head, it disappears, along with the helmet.

I'm sure it's not the clipping field, because they disappear suddenly, from one frame to another.

I'm also sure this bug is not limited to @donmccurdy 's gltf viewer ( https://gltf-viewer.donmccurdy.com/ ), as I tested it in a project made from scratch recently, and the bug appears there, too.

It works flawlessly in the babylon sandbox: http://sandbox.babylonjs.com/, so I believe this has something to do with three.js.

Also, I think it has something to do with the fact that the mesh is skinned. If I apply the armature pose in blender before exporting, then the model appears correctly in three.js as well.

I attach a zip with 2 model files, one with skin, the other one with the applied pose:
[player.zip](https://github.com/mrdoob/three.js/files/2205971/player.zip)

Three.js:
![capture2](https://user-images.githubusercontent.com/10392261/42885336-02355ef4-8aa1-11e8-897b-531b6a133272.PNG)

Babylon.js:
![capture](https://user-images.githubusercontent.com/10392261/42885318-f717b404-8aa0-11e8-8611-1cd9f75d7a1c.PNG)

##### Three.js version

- [ ] Dev
- [x] r94
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

**Comment by looeee** [comment]
Tested it in my [loader](https://blackthread.io/loader/) and it's working fine. So whatever the issue is, it's not a three.js bug.

**Comment by soadzoor** [comment]
Yes, right. I wouldn't call it "clipping" though, but "disappearing", since the whole "mesh" disappears suddenly.

I also had the head and the helmet joined together with the rest of the body, so they would be one single mesh, but that didn't really make any difference, probably because it has a different material, so it becomes a different mesh in three.js. As far as I know, in glTF, it always becomes a separate mesh, when it has different materials, at least in three.js

**Comment by donmccurdy** [comment]
> As far as I know, in glTF, it always becomes a separate mesh, when it has different materials, at least in three.js

This depends on how the exporter creates the file — three.js does support multi-material meshes, but the exporter might not write data that way, or there might be some technical reason the exporter indicates a multi-material mesh but three.js has to split it (e.g. we can't have one mesh containing both points and triangles).

**Comment by soadzoor** [comment]
@WestLangley yes, `frustumCulled = false` works, I can confirm. But I don't understand. Skinning can translate the geometry, that's ok, but the bounding boxes should be translated accordingly, shouldn't they?

**Comment by WestLangley** [comment]
Bounding boxes exist only on the CPU, and that is where the frustum culling is occurring. Skinning occurs on the GPU.

So in your case, either the skinned vertices are translated away from the bounding box, or the bounding box is not correct for some reason. See if you can track that down.

**Comment by soadzoor** [comment]
I see. Thanks!

**Comment by WestLangley** [comment]
You can enlarge the head's bound box, manually, or perhaps set it to be the union of the bounding boxes of the head and body.
```javascript
boundingBox.union( boundingBox2 );
```
Whatever you do, do it just once, not every frame.

**Comment by soadzoor** [comment]
@WestLangley thanks for the tip, seem like a pretty nice workaround. I'll close the thread I guess

**Comment by looeee** [comment]
I've been testing this in quite a few models, and it seems like every skinned model that consists of multiple pieces exhibits this behaviour. Testing the same models in other system (Babylon.js, FBX review), they display correctly. So I think we do need to consider this a three.js bug. 

Here's a video of the most obvious example I could find (model was shared in #14520). Kind of NSFW I guess. 
[skin_bug.zip](https://github.com/mrdoob/three.js/files/2223364/skin_bug.zip)

**Comment by benolayinka** [comment]
Has there been any progress on this? I have a scene with multiple animated characters, and the bounding boxes for the skinned meshes are way off the models.

<img width="612" alt="Screenshot 2020-03-20 at 11 39 36" src="https://user-images.githubusercontent.com/6963436/77156711-35507b80-6aa0-11ea-861e-b636e3691e45.png">

Maybe unrelated, but I also noticed that the way the animated models are created in Blender, I end up with an "Object3D" parent that has a Skeleton child and a SkinnedMesh child. [SkinnedMesh ](https://threejs.org/docs/#api/en/objects/SkinnedMesh) expects to have a Skeleton child, but my animations are working so maybe this doesn't matter anyway. I only noticed because I can't add a SkeletonHelper to the SkinnedMesh - I have to create it from the Parent.

<img width="208" alt="Screenshot 2020-03-20 at 11 52 26" src="https://user-images.githubusercontent.com/6963436/77157403-9036a280-6aa1-11ea-87f2-4c828c9571c1.png">
<img width="255" alt="Screenshot 2020-03-20 at 11 51 05" src="https://user-images.githubusercontent.com/6963436/77157408-9462c000-6aa1-11ea-8a30-fb926b4d1302.png">

**Comment by donmccurdy** [comment]
>  I also noticed that the way the animated models are created in Blender, I end up with an "Object3D" parent that has a Skeleton child and a SkinnedMesh child.

This is just how the Blender exporter creates models.

SkinnedMesh expects to have a Skeleton child, but my animations are working so maybe this doesn't matter anyway.

SkinnedMesh doesn't depend on its bones being a child; it is common but not a requirement.

> I only noticed because I can't add a SkeletonHelper to the SkinnedMesh - I have to create it from the Parent.

I've been doing:

```js
const helper = new SkeletonHelper( mesh.skeleton.bones[ 0 ].parent );
```

This should work regardless of whether the mesh is the parent or not.

**Comment by iar5** [comment]
Using three^0.124.0 and Blender 2.9
In case someone else is running into this problem I want to share my workaround to fix this. 

`    let mesh = gltf.scene.children[0]
    let geometry = mesh.children[1].geometry
    let m = new THREE.Matrix4()
    m.makeRotationX(THREE.MathUtils.degToRad(90))
    m.scale(new THREE.Vector3(100, 100, 100))
    geo.boundingBox.applyMatrix4(m)
    geo.boundingSphere.applyMatrix4(m)
 `

**Comment by hybridherbst** [comment]
I don't have control over threejs in most cases, e.g. when using model-viewer or other existing viewers, so code workarounds are not possible.

For reference, @mrdoob had mentioned here

- google/model-viewer#3189 (comment)

that he came up with an efficient solution; but I'm not sure that's already implemented anywhere?

**Comment by Mugen87** [comment]
Related #21507.

**Comment by puxiao** [comment]
I encountered the same problem:

1. Some SkinnedMesh will disappear inexplicably when orbitControls changes.

   > Letting all skinnedMesh .frustumCulled = false is the only solution found that works so far.

2. Unable to correctly obtain Box3 of gltf.scene

   > I tried traversing and merging the boundingBox of all skinnedMesh, but the result was not as expected


Is there currently a correct solution?

**Comment by Mugen87** [comment]
> Unable to correctly obtain Box3 of gltf.scene

Make sure you call `updateMatrixWorld()` before computing the bounding box. It does not work without up-to-date world matrices.

**Comment by puxiao** [comment]
> > Unable to correctly obtain Box3 of gltf.scene
> 
> Make sure you call `gltf.scene.updateMatrixWorld()` before computing the bounding box. It does not work without up-to-date world matrices.

It really works for me. Thanks.

**Comment by Mugen87** [comment]
There is actually issues tracking the missing world matrix computation in `Box3`: #21404

I have tried to fix this issue in #21411 since developers frequently report incorrect bounding volumes but unfortunately the PR did not get enough support. 

@donmccurdy I have noticed that many users struggle with `Box3` since they compute bounding volumes in the `onLoad()` callback of `GLTFLoader` for `gltf.scene`. We could mitigate these issues by calling `gltf.scene.updateMatrixWorld()` inside the loader right before the `onLoad()` callback fires. If we do not update matrices in `Box3.setFromObject()`, can we at least ensure computed world matrices in the `onLoad()` callback?
